### ISSUE TITLE
onAuthStateUiChange(authStateHandler) can call authStateHandler after unsubscribe with AuthState changes to SignedIn 

### ISSUE THREAD

**Issue by bwindsor** [aws-amplify/amplify-js#7613]
**Describe the bug**
It is possible for the `authStateHandler` to be called after the unsubscribe function is called. In particular, this happens when `authState` changes to `AuthState.SignedIn`.

**To Reproduce**
1. Create an auth backend and create a test Cognito user.
1. Create a file with the **code snippet** below in it.
1. Render `MyComponent` in your app.
1. Open your app in the browser and open dev tools
1. Enter the username and password of the test user and click **Sign In**
1. Observe that the change handler is called once more after the `unsubscribe` function has been called

**Expected behavior**
I expect that `authStateHandler` is never called after `unsubscribe` is called.

**Code Snippet**
```ts
const MyComponent = () => {
    const [signedIn, setSignedIn] = React.useState<boolean>(false);

    return signedIn ? <div>You are signed in</div> : <MyAuthenticator onSignedIn={() => setSignedIn(true)}/>;
}

type MyAuthenticatorProps = {
    onSignedIn: () => void;
}
const MyAuthenticator: React.FC<MyAuthenticatorProps> = ({onSignedIn}) => {
    React.useEffect(() => {
        console.log('subscribe');
        const authStateHandler: AuthStateHandler = (nextAuthState, authData) => {
            console.log('change handler called: ' + nextAuthState);
            if (nextAuthState === AuthState.SignedIn) {
                onSignedIn();
            }
        }
        const unsubscribe = onAuthUIStateChange(authStateHandler);        return () => {
            console.log('unsubscribe');
            unsubscribe();
        }
    }, [onSignedIn]);

    return <AmplifyAuthenticator />
}
```

**Screenshots**
Observed console output. For my test user pool I had MFA enabled, hence the `confirmSignIn` step, but whether that is there or not is irrelevant to this issue.
![image](https://user-images.githubusercontent.com/22724286/105639504-9f7c3e80-5e70-11eb-9683-0b7f055f65d7.png)

**What is Configured?**
My Cognito pool is created by terraform. I am only using Amplify for the UI, not to configure backend resources.
I configure Amplify as follows
```ts
import { Auth } from '@aws-amplify/auth';
Auth.configure({
    // REQUIRED - Amazon Cognito Region
    region: config.AWS_REGION,
    // OPTIONAL - Amazon Cognito User Pool ID
    userPoolId: config.USER_POOL_ID,
    // OPTIONAL - Amazon Cognito Web Client ID (26-char alphanumeric string)
    userPoolWebClientId: config.USER_POOL_WEB_CLIENT_ID,
});
```

<details>
  <summary><strong>Environment</strong></summary>

```
  System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Memory: 6.27 GB / 15.86 GB
  Binaries:
    Node: 14.15.4 - C:\Program Files\nodejs\node.EXE
    npm: 7.3.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 87.0.4280.141
    Edge: Spartan (44.18362.449.0)
    Internet Explorer: 11.0.18362.1
  npmPackages:
    @aws-amplify/auth: ^3.4.18-unstable.6 => 3.4.18-unstable.6+286c9e8fb
    @aws-amplify/ui-react: ^0.2.35-unstable.6 => 0.2.35-unstable.6+286c9e8fb
    @types/history: ^4.7.8 => 4.7.8
    @types/lodash: ^4.14.168 => 4.14.168
    @types/react-dom: ^16.9.10 => 16.9.10
    @types/react-router-dom: ^5.1.6 => 5.1.7
    chart.js: ^2.9.4 => 2.9.4
    history: ^4.10.1 => 4.10.1
    http-proxy-middleware: ^1.0.6 => 1.0.6
    react: ^17.0.1 => 17.0.1
    react-admin: ^3.11.2 => 3.11.2
    react-chartjs-2: ^2.11.1 => 2.11.1
    react-dom: ^17.0.1 => 17.0.1
    react-scripts: ^3.4.4 => 3.4.4
    typescript: ^4.1.0-beta => 4.1.0-dev.20201027
  npmGlobalPackages:
    create-next-app: 10.0.5
    create-react-app: 4.0.1
    create-react-native-module: 0.16.0
    http-server-spa: 1.3.0
    http-server: 0.11.1
    jetifier: 1.6.6
    npm-force-resolutions: 0.0.3
    npm-windows-upgrade: 6.0.1
    npx: 10.2.2
    react-native: 0.62.2
    serverless: 1.83.2
```

</details>

**Smartphone (please complete the following information):**
N/A

**Additional context**
It looks like the reason that this occurs is apparent in the amplify source here: https://github.com/aws-amplify/amplify-js/blob/b5de2c99a6aafbafe961849c8066631c2a03151a/packages/amplify-ui-components/src/common/helpers.ts#L67-L68

Because of the async nature of calling `currentAuthenticatedUser`, if `unsubscribe` is called after `currentAuthenticatedUser` is called but before the promise resolves, `authStateHandler` will be called after `unsubscribe`.

**Comment by jacoblogan** [comment]
Thank you for your detailed and insightful report, the issue has been reproduced and is being worked on.

**Comment by wlee221** [comment]
Hi @bwindsor, we merged #8428 that resolves this issue. This will be published in the next release cycle. Please let us know if the issue persists after then.
